posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,218 @@
+---
+author: 'Yaël Dillies, Paul Lezeau'
+category: 'Metaprogramming'
+date: 2025-04-16 12:00:00 UTC+01:00
+description: 'Informal introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: fantastic-simprocs
+tags: 'simp, simproc, meta'
+title: Fantastic Simprocs and Where to Find Them.
+type: text
+---
+
+Lean core added support for custom simplification procedures, aka *simprocs*.
+This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+This post describes purely informally what simprocs are and do.
+The second post will be a walkthrough to writing a simple simproc in three different ways.
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs.
+
+# How simp works
+
+`simp` is made of two components.
+
+The first component is **rewriting rules**.
+Almost all rewriting rules are lemmas that prove an equality and are tagged with `@[simp]` in eg Lean or Mathlib.
+A rewriting rule is characterised by its *left hand side* and *right hand side*.
+Eg for a lemma of the form `LHS = RHS`, this is `LHS` and `RHS` respectively.
+
+The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
+When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.
+
+For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+> The order in which `simp` traverses an expression is relatively intricate.
+> As a simple approximation, simplification is performed from the inside-out.
+> See the next blog post for full details on the algorithm.
+
+> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.
+
+# What is a simproc?
+
+Earlier, we said "almost all rewriting rules are lemmas".
+What about rewriting rules that are not lemmas?
+This is where simprocs enter the picture.
```

Comment:
Where do theorems like `: Even 0` fit into this narrative? Perhaps worth a footnote that theorem of the form `P` are often treated as `P ↔ True`?

Reply:
I've added an explanation earlier